src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/movements/MovementsService.kt

```diff
@@ -477,6 +494,66 @@ class MovementsService(
   private fun addressOrThrow(addressId: Long) = addressRepository.findByIdOrNull(addressId)
     ?: throw BadDataException("Address id $addressId is invalid")
 
+  private fun findOrCreateAddress(request: UpsertTemporaryAbsenceAddress, offender: Offender): Address {
+    // If an id is passed then we don't need to create an address
+    if (request.id != null) return addressOrThrow(request.id)
+
+    if (request.addressText == null || request.ownerClass == null) throw BadDataException("Both address text and owner class are required to create a new address")
+    if (request.ownerClass in listOf("AGY", "CORP") && request.name == null) throw BadDataException("Corporate or Agency name is required to create a new address")
+
+    return when (request.ownerClass) {
+      "OFF" -> createOffenderAddress(request.addressText, request.postalCode, offender)
+      // Agency maintenance is tightly controlled in NOMIS so we'll always create a corporate address (which is the same as users often do now)
+      "AGY", "CORP" -> createCorporateAddress(request.name!!, request.addressText, request.postalCode)
+      else -> throw BadDataException("Address owner class ${request.ownerClass} is not supported")
+    }
+  }
+
+  private fun formatAddressText(addressText: String): Pair<String, String?> {
+    val maxPremiseLength = 135
+
+    if (addressText.length <= maxPremiseLength) {
+      return addressText.trim() to null
+    }
+
+    val split = if (addressText.substring(maxPremiseLength, maxPremiseLength + 1) == " ") {
+      maxPremiseLength
+    } else {
+      addressText.substring(0, maxPremiseLength).lastIndexOf(" ")
+    }
+    return addressText.substring(0, split).trim() to addressText.substring(split, addressText.length).trim()
+  }
+
+  private fun createOffenderAddress(addressText: String, postalCode: String?, offender: Offender): OffenderAddress {
+    val (premise, street) = formatAddressText(addressText)
```

Comment:
Rather than attempting to split the address into its components (street, area etc.) we're just saving the entire text in column PREMISE and overflowing into column STREET if too long. (that gives us 295 characters which seems to be more than enough).